### PR TITLE
adding qualified name lookup for psci :browse issue #974

### PR DESCRIPTION
**Pull request by soupi**
as requested [here](https://github.com/purescript/purescript/issues/974).

### CODE REVIEW

File: psci/PSCi.hs [hunk 1]
```diff
@@ -487,7 +487,11 @@ handleBrowse moduleName = do
     Left errs -> printErrors errs
     Right env' ->
       if moduleName `notElem` (nub . map ((\ (P.Module _ _ modName _ _ ) -> modName) . snd)) (psciLoadedModules st)
-        then PSCI $ outputStrLn $ "Module '" ++ N.runModuleName moduleName ++ "' is not valid."
+        then case Just moduleName `lookup` (nub . map ( \(modName, _, mayQuaName) -> (mayQuaName, modName))) (psciImportedModules st) of
+          Nothing ->
+            PSCI $ outputStrLn $ "Module '" ++ N.runModuleName moduleName ++ "' is not valid."
+          Just qualifiedName ->
+            handleBrowse qualifiedName
```
Comment (paf31) [inline psci/PSCi.hs]: Here, we're running `make` twice. Can we avoid that?

Comment (soupi) [inline psci/PSCi.hs]: We can just copy and run the `if` on line 489 again instead, would that be better?

Comment (paf31) [inline psci/PSCi.hs]: Well, let's not copy it. Maybe we can extract it into a separate function.

File: psci/PSCi.hs [hunk 2]
```diff
@@ -486,9 +486,24 @@ handleBrowse moduleName = do
...
+        else case lookupUnQualifiedModName moduleName st of
+          Just unQualifiedName ->
+            if isModInEnv unQualifiedName st
+              then printModSigs unQualifiedName env'
+              else failNotInEnv moduleName
+          Nothing ->
+            failNotInEnv moduleName
+  where
+    isModInEnv modName st =
+        modName `elem` (nub . map ((\ (P.Module _ _ mName _ _ ) -> mName) . snd)) (psciLoadedModules st)
```
Comment (paf31) [inline psci/PSCi.hs]: Perhaps use `find` instead of `lookup` with `map`. The `nub` also seems redundant.

File: psci/PSCi.hs [hunk 3]
```diff
@@ -486,9 +486,24 @@ handleBrowse moduleName = do
   case env of
     Left errs -> printErrors errs
     Right env' ->
-      if moduleName `notElem` (nub . map ((\ (P.Module _ _ modName _ _ ) -> modName) . snd)) (psciLoadedModules st)
-        then PSCI $ outputStrLn $ "Module '" ++ N.runModuleName moduleName ++ "' is not valid."
-        else printModuleSignatures moduleName env'
+      if isModInEnv moduleName st
+        then printModSigs moduleName env'
+        else case lookupUnQualifiedModName moduleName st of
```
Comment (paf31) [inline psci/PSCi.hs]: Much nicer, thanks!

**Comment by paf31** [comment]
@garyb Let's include this one in 0.8, yes?

**Comment by garyb** [comment]
Sure, sounds good to me :+1:

File: psci/PSCi.hs [hunk 4]
```diff
@@ -486,9 +486,24 @@ handleBrowse moduleName = do
...
+          Just unQualifiedName ->
+            if isModInEnv unQualifiedName st
+              then printModSigs unQualifiedName env'
+              else failNotInEnv moduleName
+          Nothing ->
+            failNotInEnv moduleName
+  where
+    isModInEnv modName st =
+        isJust $ find ((\ (P.Module _ _ mName _ _ ) -> mName == modName) . snd) (psciLoadedModules st)
```
Comment (paf31) [inline psci/PSCi.hs]: `isJust . find = any`

Comment (paf31) [inline psci/PSCi.hs]: Also, you can then eta-reduce:

```
isModInEnv modName = any ((== modName) . getModuleName . snd) . psciLoadedModules
```

File: psci/PSCi.hs [hunk 5]
```diff
@@ -486,9 +486,24 @@ handleBrowse moduleName = do
...
+    failNotInEnv modName =
+        PSCI $ outputStrLn $ "Module '" ++ N.runModuleName modName ++ "' is not valid."
+    printModSigs modName env' =
```
Comment (paf31) [inline psci/PSCi.hs]: `printModSigs = printModuleSignatures`

**Comment by paf31** [comment]
@soupi I added a couple more comments here. If you could have a quick look, I'll get this merged in before the next RC. Thanks!

**Comment by soupi** [comment]
@paf31 thanks for the suggestions! I made the changes and squashed the commits by doing a force push. I hope that's alright.

**Comment by paf31** [comment]
Thanks very much!

**Comment by soupi** [comment]
gladly!

**Pull request merged**